gammapy/irf/core.py

```diff
@@ -1,17 +1,284 @@
+            Data array
+        """
+        required_shape = self.axes.shape
 
-class IRFMap:
-    """IRF map base class"""
-    _axis_names = []
+        if np.isscalar(value):
+            value = value * np.ones(required_shape)
+
+        if isinstance(value, u.Quantity):
+            raise TypeError("Map data must be a Numpy array. Set unit separately")
+
+        if np.shape(value) != required_shape:
+            raise ValueError(
+                f"data shape {value.shape} does not match"
+                f"axes shape {required_shape}"
+            )
+
+        self._data = value
+
+        # reset cached interpolators
+        self.__dict__.pop("_interpolate", None)
+        self.__dict__.pop("_integrate_rad", None)
+
+    @lazyproperty
+    def _interpolate(self):
+        points = [a.center for a in self.axes]
+        points_scale = tuple([a.interp for a in self.axes])
+        return ScaledRegularGridInterpolator(
+            points, self.quantity, points_scale=points_scale, **self.default_interp_kwargs
+        )
+
+    @property
+    def quantity(self):
+        """`~astropy.units.Quantity`"""
+        return u.Quantity(self.data, unit=self.unit, copy=False)
+
+    @quantity.setter
+    def quantity(self, val):
+        val = u.Quantity(val, copy=False)
+        self.data = val.value
+        self.unit = val.unit
+
+    @property
+    def axes(self):
+        """`MapAxes`"""
+        return self._axes
+
+    def __str__(self):
+        str_ = f"{self.__class__.__name__}\n"
+        str_ += "-" * len(self.__class__.__name__) + "\n\n"
+        str_ += f"\taxes  : {self.axes.names}\n"
+        str_ += f"\tshape : {self.data.shape}\n"
+        str_ += f"\tndim  : {len(self.axes)}\n"
+        str_ += f"\tunit  : {self.unit}\n"
+        str_ += f"\tdtype : {self.data.dtype}\n"
+        return str_.expandtabs(tabsize=2)
+
+    def evaluate(self, method=None, **kwargs):
+        """Evaluate IRF
+
+        Parameters
+        ----------
+        **kwargs : dict
+            Coordinates at which to evaluate the IRF
+        method : str {'linear', 'nearest'}, optional
+            Interpolation method
+
+        Returns
+        -------
+        array : `~astropy.units.Quantity`
+            Interpolated values
+        """
+        # TODO: change to coord dict?
+        non_valid_axis = set(kwargs).difference(self.axes.names)
+        if non_valid_axis:
+            raise ValueError(
+                f"Not a valid coordinate axis {non_valid_axis}"
+                f" Choose from: {self.axes.names}"
+            )
+
+        coords_default = self.axes.get_coord()
+
+        for key, value in kwargs.items():
+            coord = kwargs.get(key, value)
+            if coord is not None:
+                coords_default[key] = u.Quantity(coord, copy=False)
+
+        return self._interpolate(coords_default.values(), method=method)
+
+    # TODO: define a proper integration method
+    def integrate_energy(self, method="linear", **kwargs):
```

Comment:
Can integrate methods be general enough that they belong to the more abstract base class?
Here the integration is on reco energy, hence the method is only applicable to background. (And background is the only IRF defined as a differential quantity over energy).
I would suggest to put an energy integration method only on the `Background` base class.

Reply:
Good point. I'll check whether I can make this maybe a more general integration method. Otherwise I'll move it to `Background` again...

Reply:
I renamed this method to `integrate_log_log()` and added an `axis_name` argument. Maybe it's useful in another context as well.